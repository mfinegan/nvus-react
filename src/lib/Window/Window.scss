@import '../theme';

.window-container {
    border-bottom-left-radius: $window-bottom-radius;
    border-bottom-right-radius: $window-bottom-radius;
    background-color: $window-bg-color;
    margin-left: 0.5rem;
    margin-right: 0.5rem;
    height: 100%;
    width: 100%;

    .window-header {
        font-family: $window-header-text;
        background-color: $window-header-bg-color;
        display: flex;

        .window-title {
            margin-top: auto;
            margin-bottom: auto;
            margin-left: 0.5rem;
            font-weight: lighter;
            color: $text-color;
            display: flex;
            align-items: center;

            .window-title-label {
                margin-left: 0.75rem;
            }
        }

        .window-header-buttons {
            height: $window-header-buttons-height;
            width: 5rem;
            border-radius: 0;
            display: flex;
            margin-left: auto;

            .window-header-button {
                cursor: pointer;
                margin: auto;
                height: inherit;
                width: $window-header-buttons-height;
                justify-content: center;
                background-color: $window-header-buttons-bg-color;
                color: $text-color;
                border: none;
                transition: $window-button-transition;
                opacity: $window-header-buttons-opacity;

                &.button-disabled {
                    cursor: default;
                    background-color: $window-button-disabled-bg-color;
                    opacity: $window-header-button-disabled-opacity;
                    transition: $window-button-transition;
                }

                &.pinned-active {
                    opacity: 0.75;
                }

                &:hover {
                    background-color: rgba($color-secondary-2-4, 0.25);
                }
            }
        }
    }

    .window-body {
        background-color: $window-body-bg-color;
        font-family: $window-body-text;
        color: $text-color;
        justify-content: center;
        align-items: center;
        display: block;
        flex-direction: column;
        margin: auto;
        height: calc(100% - 2.4rem);
        border-bottom-left-radius: $window-bottom-radius - 0.1rem;
        border-bottom-right-radius: $window-bottom-radius - 0.1rem;
    }
}

.react-grid-placeholder {
    background-color: $color-secondary-2-4;
    border-bottom-left-radius: $window-bottom-radius;
    border-bottom-right-radius: $window-bottom-radius;
}
