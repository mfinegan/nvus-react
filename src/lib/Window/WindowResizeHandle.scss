@import '../theme';

.window-container.draggable + .window-resize-handle {
    background: transparent;
    position: absolute;
    bottom: 0.25rem;
    right: -0.25rem;
    padding: 0;
    cursor: se-resize;
    transform: scaleX(-1);
    visibility: visible;
    color: $text-color;
    opacity: 0.3;
    transition: $window-button-transition;
}

.window-container + .window-resize-handle {
    visibility: hidden;
    opacity: 0.01;
    transition: $window-button-transition;
}

.react-grid-item > .react-resizable-handle.react-resizable-handle-se {
    bottom: 0.25rem;
    right: -0.25rem;
}

.react-resizable-handle {
    background: transparent;
}
