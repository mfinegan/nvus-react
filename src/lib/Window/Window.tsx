import React, { ReactNode, useEffect} from "react";
import './Window.scss';
import UseWindow from './Hooks/useWindow';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'


export type ReactGridLayoutComponent = {
    i: string,
    x: number,
    y: number,
    w: number,
    h: number,
    minW?: number,
    minH?: number
    static?: boolean,
    isDraggable?: boolean,
    isResizeable?: boolean,
    child: React.ReactNode,
    icon: ReactNode
};


type WindowProps = {

    /** Title of the window */
    title: string,

    /** Child component to render */
    child?: React.ReactNode,

    /** State characteristics of this window */
    layout: ReactGridLayoutComponent,

    /** Function Callback for when Pinned Action is executed */
    onPinned: Function,

    /** Function callback for when Close Window Action is executed */
    onClosed: Function
};

function Window({ title, child, layout, onPinned, onClosed }: WindowProps) {
  
    const { isOpen, isPinned, toggleOpen, togglePinned} = UseWindow({key:layout.i});

    useEffect(()=>{
        if(isPinned)
            onPinned(layout.i);
    },[isPinned, layout.i, onPinned]);

    useEffect(()=>{
        if(!isOpen)
            onClosed(layout.i);
    },[isOpen, layout.i, onClosed]);

    const handleClick = (evt: React.MouseEvent<HTMLButtonElement, MouseEvent>) => { evt.stopPropagation(); };

    return (

        <div className={!isPinned ? "window-container draggable " : " window-container"}>
            <div className="window-header p-d-flex">

                <div className="window-title">
                    <div className="icon-container">{layout.icon} </div>
                    <span className="window-title-label">{title.toUpperCase()}</span>
                </div>
                <div className="window-header-buttons ">
                    <button className={isPinned ? "window-header-button pinned-active" : "window-header-button"} onClick={togglePinned} onMouseDown={e => handleClick(e)}>
                        <FontAwesomeIcon icon={["fad", "thumbtack"]} />
                    </button>
                    <button className={!isPinned? "window-header-button":"window-header-button button-disabled"} onClick={toggleOpen} disabled={isPinned} onMouseDown={e => handleClick(e)}>
                        <FontAwesomeIcon icon={["fad", "window-close"]} />
                    </button>
                </div>
            </div>
            <div className="window-body">
                <div className="window-body-content">
                    {child}
                </div>
            </div>
            <div className="window-footer">
            </div>
        </div>
    );
}

export default Window;