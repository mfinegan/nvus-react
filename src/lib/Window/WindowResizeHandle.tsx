import React from 'react'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faExpandAlt } from '@fortawesome/free-solid-svg-icons'
import './WindowResizeHandle.scss'

/**
 * Wrapper component for the resize icon to be used as a resize handle
 */

const BottomRightHandle = React.forwardRef<HTMLDivElement>((props, ref) => {
    return (
        <div {...props} ref={ref} className={'window-resize-handle'}>
            <FontAwesomeIcon className="resize-handle-se" icon={faExpandAlt} />
        </div>
    )
})

BottomRightHandle.displayName = 'BottomRightHandle'

export default BottomRightHandle
