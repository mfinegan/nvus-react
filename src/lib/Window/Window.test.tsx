import React from 'react';
import { renderHook, act } from '@testing-library/react-hooks'
import UseWindow from './Hooks/useWindow'
import Window from './Window'
import { shallow } from 'enzyme'

test('renders Window with passed props', () => {
    const { result } = renderHook(() => UseWindow({ key: 'Test' }))

    expect(result.current.isPinned).toBe(false)
    expect(result.current.isOpen).toBe(true)
})

test('Set Window Pinned Status to True When TogglePinned called', () => {
    const { result } = renderHook(() => UseWindow({ key: 'Test' }))

    expect(result.current.isPinned).toBe(false)

    act(() => {
        result.current.togglePinned()
    })

    expect(result.current.isPinned).toBe(true)
    expect(result.current.isOpen).toBe(true)
})

test('Verify toggleOpen does nothing if isPinned is true', () => {
    const { result } = renderHook(() => UseWindow({ key: 'Test' }))

    expect(result.current.isPinned).toBe(false)

    act(() => {
        result.current.togglePinned()
    })

    expect(result.current.isPinned).toBe(true)
    expect(result.current.isOpen).toBe(true)

    act(() => {
        result.current.toggleOpen()
    })

    expect(result.current.isPinned).toBe(true)
    expect(result.current.isOpen).toBe(true)
})

test('Set Open Status to False When ToggleOpen called', () => {
    const { result } = renderHook(() => UseWindow({ key: 'Test' }))

    expect(result.current.isPinned).toBe(false)
    expect(result.current.isOpen).toBe(true)

    act(() => {
        result.current.toggleOpen()
    })

    expect(result.current.isPinned).toBe(false)
    expect(result.current.isOpen).toBe(false)
})

test('Render component Test', () => {
    const el = {
        i: 'Test Title',
        x: 1,
        y: 1,
        w: 4,
        h: 4,
        minW: 4,
        minH: 4,
        child: <span>Test</span>,
    }
    const comp = shallow(
        <Window
            title={el.i}
            child={el.child}
            layout={el}
            onPinned={() => alert('pinned')}
            onClosed={() => alert('closed')}
        />
    )

    expect(comp).toMatchSnapshot()
})
