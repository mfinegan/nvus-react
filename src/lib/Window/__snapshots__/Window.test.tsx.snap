// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render component Test 1`] = `
<div
  className="window-container draggable "
>
  <div
    className="window-header p-d-flex"
  >
    <div
      className="window-title"
    >
      <div
        className="icon-container"
      >
         
      </div>
      <span
        className="window-title-label"
      >
        TEST TITLE
      </span>
    </div>
    <div
      className="window-header-buttons "
    >
      <button
        className="window-header-button"
        onClick={[Function]}
        onMouseDown={[Function]}
      >
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={false}
          flip={null}
          icon={
            Object {
              "icon": Array [
                384,
                512,
                Array [],
                "f08d",
                "M298.028 214.267L285.793 96H328c13.255 0 24-10.745 24-24V24c0-13.255-10.745-24-24-24H56C42.745 0 32 10.745 32 24v48c0 13.255 10.745 24 24 24h42.207L85.972 214.267C37.465 236.82 0 277.261 0 328c0 13.255 10.745 24 24 24h136v104.007c0 1.242.289 2.467.845 3.578l24 48c2.941 5.882 11.364 5.893 14.311 0l24-48a8.008 8.008 0 0 0 .845-3.578V352h136c13.255 0 24-10.745 24-24-.001-51.183-37.983-91.42-85.973-113.733z",
              ],
              "iconName": "thumbtack",
              "prefix": "fas",
            }
          }
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          swapOpacity={false}
          symbol={false}
          title=""
          transform={null}
        />
      </button>
      <button
        className="window-header-button"
        disabled={false}
        onClick={[Function]}
        onMouseDown={[Function]}
      >
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={false}
          flip={null}
          icon={
            Object {
              "icon": Array [
                512,
                512,
                Array [],
                "f410",
                "M464 32H48C21.5 32 0 53.5 0 80v352c0 26.5 21.5 48 48 48h416c26.5 0 48-21.5 48-48V80c0-26.5-21.5-48-48-48zm-83.6 290.5c4.8 4.8 4.8 12.6 0 17.4l-40.5 40.5c-4.8 4.8-12.6 4.8-17.4 0L256 313.3l-66.5 67.1c-4.8 4.8-12.6 4.8-17.4 0l-40.5-40.5c-4.8-4.8-4.8-12.6 0-17.4l67.1-66.5-67.1-66.5c-4.8-4.8-4.8-12.6 0-17.4l40.5-40.5c4.8-4.8 12.6-4.8 17.4 0l66.5 67.1 66.5-67.1c4.8-4.8 12.6-4.8 17.4 0l40.5 40.5c4.8 4.8 4.8 12.6 0 17.4L313.3 256l67.1 66.5z",
              ],
              "iconName": "window-close",
              "prefix": "fas",
            }
          }
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          swapOpacity={false}
          symbol={false}
          title=""
          transform={null}
        />
      </button>
    </div>
  </div>
  <div
    className="window-body"
  >
    <div
      className="window-body-content"
    >
      <span>
        Test
      </span>
    </div>
  </div>
  <div
    className="window-footer"
  />
</div>
`;
